File: ruby/lib/spanvalue/proto.rb
# frozen_string_literal: true

module Spanvalue
  module Proto
    module_function

    def get(obj, *names, default: nil)
      return default if obj.nil?

      if obj.is_a?(Hash)
        names.each do |name|
          key = name.is_a?(Symbol) ? name : name.to_s
          return obj[key] if obj.key?(key)
          camel = camelize(name.to_s)
          return obj[camel] if obj.key?(camel)
        end
        return default
      end

      names.each do |name|
        return obj.public_send(name) if obj.respond_to?(name)
      end

      if obj.respond_to?(:[])
        names.each do |name|
          val = obj[name]
          return val unless val.nil?
        end
      end

      default
    end

    def camelize(s)
      parts = s.split('_')
      parts[0] + parts[1..].map(&:capitalize).join
    end
    private_class_method :camelize

    def type_code(typ)
      Codes.parse_type_code(get(typ, :code, :Code))
    end

    def type_annotation(typ)
      Codes.parse_type_annotation(get(typ, :type_annotation, :typeAnnotation, :TypeAnnotation))
    end

    def proto_type_fqn(typ)
      get(typ, :proto_type_fqn, :protoTypeFqn, :ProtoTypeFqn, default: '') || ''
    end

    def array_element_type(typ)
      get(typ, :array_element_type, :arrayElementType, :ArrayElementType)
    end

    def struct_type(typ)
      get(typ, :struct_type, :structType, :StructType)
    end

    def struct_fields(typ)
      st = struct_type(typ)
      return [] if st.nil?

      fields = get(st, :fields, :Fields, default: [])
      fields.nil? ? [] : fields.to_a
    end

    def field_name(field)
      get(field, :name, :Name, default: '') || ''
    end

    def field_type(field)
      get(field, :type, :Type)
    end

    def value_kind(value)
      return 'null' if value.nil?

      if value.is_a?(Hash)
        return 'null' if value.key?('null_value') || value.key?('nullValue')
        return 'bool' if value.key?('bool_value') || value.key?('boolValue')
        return 'number' if value.key?('number_value') || value.key?('numberValue')
        return 'string' if value.key?('string_value') || value.key?('stringValue')
        return 'list' if value.key?('list_value') || value.key?('listValue')

        return 'missing'
      end

      return 'bool' if value.is_a?(TrueClass) || value.is_a?(FalseClass)
      return 'number' if value.is_a?(Numeric)
      return 'string' if value.is_a?(String)
      return 'list' if value.is_a?(Array)

      if value.respond_to?(:WhichOneof)
        which = value.WhichOneof('kind')
        return 'missing' if which.nil?

        case which
        when 'null_value', 'nullValue' then return 'null'
        when 'bool_value', 'boolValue' then return 'bool'
        when 'number_value', 'numberValue' then return 'number'
        when 'string_value', 'stringValue' then return 'string'
        when 'list_value', 'listValue' then return 'list'
        end
      end

      %i[bool_value boolValue number_value numberValue string_value stringValue list_value listValue].each do |attr|
        next unless value.respond_to?(attr)

        val = value.public_send(attr)
        kind = attr.to_s.sub(/_?value$/i, '').downcase
        kind = 'bool' if kind == 'bool'
        kind = 'number' if kind == 'number'
        kind = 'string' if kind == 'string'
        kind = 'list' if kind == 'list'
        return kind unless val.nil?
      end

      'missing'
    end

    def null_value?(value)
      %w[null missing].include?(value_kind(value))
    end

    def bool_value(value)
      return value if value.is_a?(TrueClass) || value.is_a?(FalseClass)
      return !!get(value, :bool_value, :boolValue) if value.is_a?(Hash)

      !!get(value, :bool_value, :boolValue)
    end

    def number_value(value)
      return value.to_f if value.is_a?(Numeric)
      return get(value, :number_value, :numberValue).to_f if value.is_a?(Hash)

      get(value, :number_value, :numberValue).to_f
    end

    def string_value(value)
      return value if value.is_a?(String)
      return get(value, :string_value, :stringValue, default: '').to_s if value.is_a?(Hash)

      get(value, :string_value, :stringValue, default: '').to_s
    end

    def list_values(value)
      return value if value.is_a?(Array)

      if value.is_a?(Hash)
        lv = value['list_value'] || value['listValue']
        if lv.is_a?(Hash)
          vals = lv['values'] || lv['Values'] || []
          return vals.to_a
        end
        return lv.values.to_a if lv.respond_to?(:values)
      end

      lv = get(value, :list_value, :listValue)
      return [] if lv.nil?

      vals = get(lv, :values, :Values, default: [])
      vals.nil? ? [] : vals.to_a
    end
  end
end

File: python/src/spanvalue/proto.py
"""Duck-typed access to protojson dicts and protobuf objects."""

from __future__ import annotations

from typing import Any

from .codes import parse_type_annotation, parse_type_code


def _get(obj: Any, *names: str, default: Any = None) -> Any:
    if obj is None:
        return default
    if isinstance(obj, dict):
        for name in names:
            if name in obj:
                return obj[name]
        return default
    for name in names:
        if hasattr(obj, name):
            return getattr(obj, name)
    getter = getattr(obj, "get", None)
    if getter is not None:
        for name in names:
            val = getter(name)
            if val is not None:
                return val
    return default


def type_code(typ: Any) -> int:
    return parse_type_code(_get(typ, "code", "Code"))


def type_annotation(typ: Any) -> int:
    return parse_type_annotation(_get(typ, "type_annotation", "typeAnnotation", "TypeAnnotation"))


def proto_type_fqn(typ: Any) -> str:
    return _get(typ, "proto_type_fqn", "protoTypeFqn", "ProtoTypeFqn", default="") or ""


def array_element_type(typ: Any) -> Any:
    return _get(typ, "array_element_type", "arrayElementType", "ArrayElementType")


def struct_type(typ: Any) -> Any:
    return _get(typ, "struct_type", "structType", "StructType")


def struct_fields(typ: Any) -> list[Any]:
    st = struct_type(typ)
    if st is None:
        return []
    fields = _get(st, "fields", "Fields", default=[])
    return list(fields) if fields is not None else []


def field_name(field: Any) -> str:
    return _get(field, "name", "Name", default="") or ""


def field_type(field: Any) -> Any:
    return _get(field, "type", "Type")


def value_kind(value: Any) -> str:
    """Return wire kind: null, bool, number, string, list, or missing."""
    if value is None:
        return "null"
    if isinstance(value, dict):
        if "null_value" in value or "nullValue" in value:
            return "null"
        if "bool_value" in value or "boolValue" in value:
            return "bool"
        if "number_value" in value or "numberValue" in value:
            return "number"
        if "string_value" in value or "stringValue" in value:
            return "string"
        if "list_value" in value or "listValue" in value:
            return "list"
        return "missing"
    if isinstance(value, bool):
        return "bool"
    if isinstance(value, (int, float)) and not isinstance(value, bool):
        return "number"
    if isinstance(value, str):
        return "string"
    if isinstance(value, list):
        return "list"
    if hasattr(value, "WhichOneof"):
        which = value.WhichOneof("kind")
        if which is None:
            return "missing"
        if which in ("null_value", "nullValue"):
            return "null"
        if which in ("bool_value", "boolValue"):
            return "bool"
        if which in ("number_value", "numberValue"):
            return "number"
        if which in ("string_value", "stringValue"):
            return "string"
        if which in ("list_value", "listValue"):
            return "list"
    for attr, kind in (
        ("bool_value", "bool"),
        ("boolValue", "bool"),
        ("number_value", "number"),
        ("numberValue", "number"),
        ("string_value", "string"),
        ("stringValue", "string"),
        ("list_value", "list"),
        ("listValue", "list"),
    ):
        if hasattr(value, attr):
            if getattr(value, attr) is not None or kind == "null":
                return kind
    return "missing"


def is_null_value(value: Any) -> bool:
    return value_kind(value) in ("null", "missing")


def bool_value(value: Any) -> bool:
    if isinstance(value, bool):
        return value
    if isinstance(value, dict):
        return bool(value.get("bool_value", value.get("boolValue")))
    return bool(_get(value, "bool_value", "boolValue"))


def number_value(value: Any) -> float:
    if isinstance(value, (int, float)) and not isinstance(value, bool):
        return float(value)
    if isinstance(value, dict):
        return float(value.get("number_value", value.get("numberValue")))
    return float(_get(value, "number_value", "numberValue"))


def string_value(value: Any) -> str:
    if isinstance(value, str):
        return value
    if isinstance(value, dict):
        return str(value.get("string_value", value.get("stringValue", "")))
    return str(_get(value, "string_value", "stringValue", default=""))


def list_values(value: Any) -> list[Any]:
    if isinstance(value, list):
        return value
    if isinstance(value, dict):
        lv = value.get("list_value", value.get("listValue"))
        if isinstance(lv, dict):
            return list(lv.get("values", lv.get("Values", [])))
        if lv is not None and hasattr(lv, "values"):
            return list(lv.values)
    lv = _get(value, "list_value", "listValue")
    if lv is not None:
        vals = _get(lv, "values", "Values", default=[])
        return list(vals) if vals is not None else []
    return []
